### PR TITLE
JSObjectReference API review follow-up

### PR DESCRIPTION
**Pull request by MackinnonBuck**
#### Description

Addresses feedback from the JSObjectReference API review (#25355).

#### Customer Impact

This is a necessary change to avoid future breaking changes to the public API.

#### Regression?

No.

#### Risk

Little risk - this is an area that is well-tested, and the updates in this PR don't make any large implementation changes.

### CODE REVIEW

File: src/JSInterop/Microsoft.JSInterop/src/IJSInProcessObjectReference.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    public interface IJSInProcessObjectReference : IJSObjectReference
+    {
+        /// <summary>
+        /// Invokes the specified JavaScript function synchronously.
+        /// </summary>
+        /// <typeparam name="TValue">The JSON-serializable return type.</typeparam>
+        /// <param name="identifier">An identifier for the function to invoke. For example, the value <c>"someScope.someFunction"</c> will invoke the function <c>someScope.someFunction</c> on the target instance.</param>
+        /// <param name="args">JSON-serializable arguments.</param>
+        /// <returns>An instance of <typeparamref name="TValue"/> obtained by JSON-deserializing the return value.</returns>
+        public TValue Invoke<TValue>(string identifier, params object?[]? args);
```
Comment (SteveSandersonMS) [inline src/JSInterop/Microsoft.JSInterop/src/IJSInProcessObjectReference.cs:18]:
```suggestion
        TValue Invoke<TValue>(string identifier, params object?[]? args);
```

File: src/JSInterop/Microsoft.JSInterop/src/IJSObjectReference.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        /// <para>
+        /// <see cref="JSRuntime"/> will apply timeouts to this operation based on the value configured in <see cref="JSRuntime.DefaultAsyncTimeout"/>. To dispatch a call with a different, or no timeout,
+        /// consider using <see cref="InvokeAsync{TValue}(string, CancellationToken, object[])" />.
+        /// </para>
+        /// </summary>
+        /// <typeparam name="TValue">The JSON-serializable return type.</typeparam>
+        /// <param name="identifier">An identifier for the function to invoke. For example, the value <c>"someScope.someFunction"</c> will invoke the function <c>someScope.someFunction</c> on the target instance.</param>
+        /// <param name="args">JSON-serializable arguments.</param>
+        /// <returns>An instance of <typeparamref name="TValue"/> obtained by JSON-deserializing the return value.</returns>
+        public ValueTask<TValue> InvokeAsync<TValue>(string identifier, object?[]? args);
```
Comment (SteveSandersonMS) [inline src/JSInterop/Microsoft.JSInterop/src/IJSObjectReference.cs:26]:
```suggestion
        ValueTask<TValue> InvokeAsync<TValue>(string identifier, object?[]? args);
```

File: src/JSInterop/Microsoft.JSInterop/src/IJSObjectReference.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        /// </summary>
+        /// <typeparam name="TValue">The JSON-serializable return type.</typeparam>
+        /// <param name="identifier">An identifier for the function to invoke. For example, the value <c>"someScope.someFunction"</c> will invoke the function <c>someScope.someFunction</c> on the target instance.</param>
+        /// <param name="cancellationToken">
+        /// A cancellation token to signal the cancellation of the operation. Specifying this parameter will override any default cancellations such as due to timeouts
+        /// (<see cref="JSRuntime.DefaultAsyncTimeout"/>) from being applied.
+        /// </param>
+        /// <param name="args">JSON-serializable arguments.</param>
+        /// <returns>An instance of <typeparamref name="TValue"/> obtained by JSON-deserializing the return value.</returns>
+        public ValueTask<TValue> InvokeAsync<TValue>(string identifier, CancellationToken cancellationToken, object?[]? args);
```
Comment (SteveSandersonMS) [inline src/JSInterop/Microsoft.JSInterop/src/IJSObjectReference.cs:39]:
```suggestion
        ValueTask<TValue> InvokeAsync<TValue>(string identifier, CancellationToken cancellationToken, object?[]? args);
```

File: src/JSInterop/Microsoft.JSInterop/src/IJSObjectReference.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.JSInterop
+{
+    /// <summary>
+    /// Represents a reference to a JavaScript object.
+    /// </summary>
+    public interface IJSObjectReference : IAsyncDisposable
```
Comment (SteveSandersonMS) [inline src/JSInterop/Microsoft.JSInterop/src/IJSObjectReference.cs:13]: Regarding nullability annotations here, whatever we do for `IJSRuntime` we should do the equivalent thing here.

We have an ongoing conversation about it here: https://github.com/dotnet/aspnetcore/pull/25261/files#r481040491

**Comment by SteveSandersonMS** [comment]
This looks really good to me. Let us know when it's ready for final review!

File: src/JSInterop/Microsoft.JSInterop/src/JSObjectReferenceExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+        /// <summary>
+        /// Invokes the specified JavaScript function asynchronously.
+        /// </summary>
+        /// <param name="jsObjectReference">The <see cref="IJSObjectReference"/>.</param>
+        /// <param name="identifier">An identifier for the function to invoke. For example, the value <c>"someScope.someFunction"</c> will invoke the function <c>someScope.someFunction</c> on the target instance.</param>
+        /// <param name="timeout">The duration after which to cancel the async operation. Overrides default timeouts (<see cref="JSRuntime.DefaultAsyncTimeout"/>).</param>
+        /// <param name="args">JSON-serializable arguments.</param>
+        /// <returns>A <see cref="ValueTask"/> that represents the asynchronous invocation operation.</returns>
+        public static async ValueTask InvokeVoidAsync(this IJSObjectReference jsObjectReference, string identifier, TimeSpan timeout, params object[] args)
```
Comment (pranavkm) [inline src/JSInterop/Microsoft.JSInterop/src/JSObjectReferenceExtensions.cs:126]:
```suggestion
        public static async ValueTask InvokeVoidAsync(this IJSObjectReference jsObjectReference, string identifier, TimeSpan timeout, params object?[] args)
```

File: src/JSInterop/Microsoft.JSInterop/src/JSObjectReferenceExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+        /// <summary>
+        /// Invokes the specified JavaScript function asynchronously.
+        /// </summary>
+        /// <param name="jsObjectReference">The <see cref="IJSObjectReference"/>.</param>
+        /// <param name="identifier">An identifier for the function to invoke. For example, the value <c>"someScope.someFunction"</c> will invoke the function <c>someScope.someFunction</c> on the target instance.</param>
+        /// <param name="timeout">The duration after which to cancel the async operation. Overrides default timeouts (<see cref="JSRuntime.DefaultAsyncTimeout"/>).</param>
+        /// <param name="args">JSON-serializable arguments.</param>
+        /// <returns>A <see cref="ValueTask"/> that represents the asynchronous invocation operation.</returns>
+        public static async ValueTask<TValue> InvokeAsync<TValue>(this IJSObjectReference jsObjectReference, string identifier, TimeSpan timeout, params object[] args)
```
Comment (pranavkm) [inline src/JSInterop/Microsoft.JSInterop/src/JSObjectReferenceExtensions.cs:105]:
```suggestion
        public static async ValueTask<TValue> InvokeAsync<TValue>(this IJSObjectReference jsObjectReference, string identifier, TimeSpan timeout, params object?[] args)
```

File: src/JSInterop/Microsoft.JSInterop/src/JSObjectReferenceExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        /// </summary>
+        /// <param name="jsObjectReference">The <see cref="IJSObjectReference"/>.</param>
+        /// <param name="identifier">An identifier for the function to invoke. For example, the value <c>"someScope.someFunction"</c> will invoke the function <c>someScope.someFunction</c> on the target instance.</param>
+        /// <param name="cancellationToken">
+        /// A cancellation token to signal the cancellation of the operation. Specifying this parameter will override any default cancellations such as due to timeouts
+        /// (<see cref="JSRuntime.DefaultAsyncTimeout"/>) from being applied.
+        /// </param>
+        /// <param name="args">JSON-serializable arguments.</param>
+        /// <returns>A <see cref="ValueTask"/> that represents the asynchronous invocation operation.</returns>
+        public static async ValueTask InvokeVoidAsync(this IJSObjectReference jsObjectReference, string identifier, CancellationToken cancellationToken, params object[] args)
```
Comment (pranavkm) [inline src/JSInterop/Microsoft.JSInterop/src/JSObjectReferenceExtensions.cs:87]:
```suggestion
        public static async ValueTask InvokeVoidAsync(this IJSObjectReference jsObjectReference, string identifier, CancellationToken cancellationToken, params object?[] args)
```

File: src/JSInterop/Microsoft.JSInterop/src/JSObjectReferenceExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+        /// <typeparam name="TValue">The JSON-serializable return type.</typeparam>
+        /// <param name="jsObjectReference">The <see cref="IJSObjectReference"/>.</param>
+        /// <param name="identifier">An identifier for the function to invoke. For example, the value <c>"someScope.someFunction"</c> will invoke the function <c>someScope.someFunction</c> on the target instance.</param>
+        /// <param name="cancellationToken">
+        /// A cancellation token to signal the cancellation of the operation. Specifying this parameter will override any default cancellations such as due to timeouts
+        /// (<see cref="JSRuntime.DefaultAsyncTimeout"/>) from being applied.
+        /// </param>
+        /// <param name="args">JSON-serializable arguments.</param>
+        /// <returns>An instance of <typeparamref name="TValue"/> obtained by JSON-deserializing the return value.</returns>
+        public static ValueTask<TValue> InvokeAsync<TValue>(this IJSObjectReference jsObjectReference, string identifier, CancellationToken cancellationToken, params object[] args)
```
Comment (pranavkm) [inline src/JSInterop/Microsoft.JSInterop/src/JSObjectReferenceExtensions.cs:66]:
```suggestion
        public static ValueTask<TValue> InvokeAsync<TValue>(this IJSObjectReference jsObjectReference, string identifier, CancellationToken cancellationToken, params object?[] args)
```

File: src/JSInterop/Microsoft.JSInterop/src/JSObjectReferenceExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+        /// <see cref="JSRuntime"/> will apply timeouts to this operation based on the value configured in <see cref="JSRuntime.DefaultAsyncTimeout"/>. To dispatch a call with a different timeout, or no timeout,
+        /// consider using <see cref="IJSObjectReference.InvokeAsync{TValue}(string, CancellationToken, object[])" />.
+        /// </para>
+        /// </summary>
+        /// <param name="jsObjectReference">The <see cref="IJSObjectReference"/>.</param>
+        /// <typeparam name="TValue">The JSON-serializable return type.</typeparam>
+        /// <param name="identifier">An identifier for the function to invoke. For example, the value <c>"someScope.someFunction"</c> will invoke the function <c>someScope.someFunction</c> on the target instance.</param>
+        /// <param name="args">JSON-serializable arguments.</param>
+        /// <returns>An instance of <typeparamref name="TValue"/> obtained by JSON-deserializing the return value.</returns>
+        public static ValueTask<TValue> InvokeAsync<TValue>(this IJSObjectReference jsObjectReference, string identifier, params object[] args)
```
Comment (pranavkm) [inline src/JSInterop/Microsoft.JSInterop/src/JSObjectReferenceExtensions.cs:44]:
```suggestion
        public static ValueTask<TValue> InvokeAsync<TValue>(this IJSObjectReference jsObjectReference, string identifier, params object?[] args)
```

File: src/JSInterop/Microsoft.JSInterop/src/JSObjectReferenceExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+    public static class JSObjectReferenceExtensions
+    {
+        /// <summary>
+        /// Invokes the specified JavaScript function asynchronously.
+        /// </summary>
+        /// <param name="jsObjectReference">The <see cref="IJSObjectReference"/>.</param>
+        /// <param name="identifier">An identifier for the function to invoke. For example, the value <c>"someScope.someFunction"</c> will invoke the function <c>someScope.someFunction</c> on the target instance.</param>
+        /// <param name="args">JSON-serializable arguments.</param>
+        /// <returns>A <see cref="ValueTask"/> that represents the asynchronous invocation operation.</returns>
+        public static async ValueTask InvokeVoidAsync(this IJSObjectReference jsObjectReference, string identifier, params object[] args)
```
Comment (pranavkm) [inline src/JSInterop/Microsoft.JSInterop/src/JSObjectReferenceExtensions.cs:22]:
```suggestion
        public static async ValueTask InvokeVoidAsync(this IJSObjectReference jsObjectReference, string identifier, params object?[] args)
```

**Review by pranavkm — APPROVED** [review]

**Comment by MackinnonBuck** [comment]
An announcement for this change was posted here: https://github.com/aspnet/Announcements/issues/435

**Comment by Pilchie** [comment]
Approved for .NET 5 RC2 - thanks for the breaking change announcement.

**Pull request merged**